### PR TITLE
Migrate provider service to state provider

### PR DESCRIPTION
**Pull request by addisonbeck**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR cuts `ProviderService` over to `StateProvider`.

## Code changes

This was a fairly painless move. Commits can be followed for each major step. The only non-standard change that needed to be made was adding a `userId` parameter to `providerService.save()` in order to facilitate logouts for account switching. 

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

## Related
* https://github.com/bitwarden/clients/pull/7895
* [AC-2010: Transition ProviderService to use StateProvider](https://bitwarden.atlassian.net/browse/AC-2010)

[AC-2010]: https://bitwarden.atlassian.net/browse/AC-2010?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/common/src/admin-console/services/provider.service.ts [hunk 1]
```diff
@@ -8,32 +10,34 @@ export const PROVIDERS = KeyDefinition.record<ProviderData>(PROVIDERS_DISK, "pro
...
+    return this.stateProvider
+      .getUserState$(PROVIDERS, userId)
+      .pipe(this.mapProviderRecordToArray());
+  }
+
+  private mapProviderRecordToArray() {
+    return map<Record<string, ProviderData>, Provider[]>((orgs) =>
+      Object.values(orgs ?? {})?.map((o) => new Provider(o)),
+    );
+  }
```
Comment (eliykat) [inline libs/common/src/admin-console/services/provider.service.ts:30]: Still references `orgs`.

**Review by eliykat — CHANGES_REQUESTED** [review]
Nicely done 👏

File: libs/common/src/state-migrations/migrations/28-move-provider-state-to-state-provider.spec.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import { ProviderMigrator } from "./28-move-provider-state-to-state-provider";
+
+function exampleProvider1() {
+  return JSON.stringify({
+    id: "id",
+    name: "name",
+    status: 0,
+    type: 0,
+    enabled: true,
+    userEvents: true,
```
Comment (justindbaur) [inline libs/common/src/state-migrations/migrations/28-move-provider-state-to-state-provider.spec.ts:15]: nit:
```suggestion
    useEvents: true,
```

**Review by justindbaur — CHANGES_REQUESTED** [review]
Just the tiniest nit and then this is good to go from platform.

File: libs/common/src/state-migrations/migrations/28-move-provider-state-to-state-provider.spec.ts (hunk 2 above)
Comment (addisonbeck) [reply libs/common/src/state-migrations/migrations/28-move-provider-state-to-state-provider.spec.ts]: Implemented from suggestion on https://github.com/bitwarden/clients/pull/8173/commits/c88324cc0f4bb82828202baa62500ed8544a1a96

File: libs/common/src/admin-console/services/provider.service.ts (hunk 1 above)
Comment (addisonbeck) [reply libs/common/src/admin-console/services/provider.service.ts]: Whoops! Changed here: https://github.com/bitwarden/clients/pull/8173/commits/28ae8e8e30c1ebdc5be06c04c94e741671f0cebb

File: libs/common/src/admin-console/services/provider.service.spec.ts [hunk 3]
```diff
@@ -21,3 +70,75 @@ describe("PROVIDERS key definition", () => {
...
+    providerService = new ProviderService(fakeStateProvider);
+  });
+
+  describe("getAll()", () => {
+    it("Returns an array of all providers stored in state", async () => {
+      const mockData: ProviderData[] = buildMockProviders(5);
+      fakeActiveUserState.nextState(arrayToRecord(mockData));
+      const providers = await providerService.getAll();
+      expect(providers).toHaveLength(5);
+      expect(providers[3]).toEqual(new Provider(mockData[3]));
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply libs/common/src/admin-console/services/provider.service.spec.ts]: Done on 
[Compare equality over entire array in a test](https://github.com/bitwarden/clients/pull/8173/commits/7cb5744b7fc27cb25cf7792b7c168658c9cb62ba)

File: apps/cli/src/bw.ts [hunk 4]
```diff
@@ -639,6 +639,7 @@ export class Main {
       this.collectionService.clear(userId as UserId),
       this.policyService.clear(userId),
       this.passwordGenerationService.clear(),
+      this.providerService.save(null, userId as UserId),
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply apps/cli/src/bw.ts:641]: I don't understand what you mean here. Does sync run for non active users? The sync seems to only run for a default user, and I was under the assumption that a sync was run on account change and then the active unlocked account was the only one ever synced.

**Review by eliykat — DISMISSED** [review]

File: apps/cli/src/bw.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply apps/cli/src/bw.ts:641]: Hmm I think you're right here, I was just on the wrong track yesterday.

**Review by justindbaur — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]
Approving merge conflict resolution on behalf of @eliykat

**Pull request merged**